test/e2e/density.go

```diff
@@ -210,31 +210,87 @@ var _ = Describe("Density", func() {
 	// Tests with "Skipped" substring in their name will be skipped when running
 	// e2e test suite without --ginkgo.focus & --ginkgo.skip flags.
 
-	for _, count := range []int{30, 50, 100} {
-		name := fmt.Sprintf("should allow starting %d pods per node", count)
-		// TODO(wojtek-t): Don't skip 30 pods per node test once #6059 if fixed.
-		if count > 0 {
-			name = "[Skipped] " + name
-		}
-		It(name, func() {
-			RCName = "my-hostname-density" + strconv.Itoa(count) + "-" + string(util.NewUUID())
-			RunRC(c, RCName, ns, "gcr.io/google_containers/pause:go", count*minionCount)
-		})
+	type Density struct {
+		skip          bool
+		totalPods     int
+		podsPerMinion int
+		rcsPerThread  int
 	}
 
-	It("[Skipped] should have master components that can handle many short-lived pods", func() {
-		threads := 5
-		var wg sync.WaitGroup
-		wg.Add(threads)
-		for i := 0; i < threads; i++ {
-			go func() {
-				defer wg.Done()
-				for i := 0; i < 10; i++ {
-					name := "my-hostname-thrash-" + string(util.NewUUID())
-					RunRC(c, name, ns, "gcr.io/google_containers/pause:go", 10*minionCount)
+	//This test should always run, even if larger densities are skipped.
+	d3 := Density{totalPods: 3, podsPerMinion: 0, rcsPerThread: 1, skip: false}
+
+	//These tests are varied and customizable.
+	//TODO (wojtek-t):don't skip d30 after #6059
+	d30 := Density{totalPods: 30, podsPerMinion: 0, rcsPerThread: 1, skip: true}
+	d50 := Density{totalPods: 50, podsPerMinion: 0, rcsPerThread: 1, skip: true}
+	d100 := Density{totalPods: 100, podsPerMinion: 0, rcsPerThread: 1, skip: true}
+	d500t5 := Density{totalPods: 500, podsPerMinion: 10, rcsPerThread: 5, skip: true}
+	d500t25 := Density{totalPods: 500, podsPerMinion: 10, rcsPerThread: 25, skip: true}
+
+	dtests := []Density{d3, d30, d50, d100, d500t5, d500t25}
+
+	//Run each test in the array which isn't skipped.
+	for i := range dtests {
+
+		//cannot do a range iterator over structs.
+		dtest := dtests[i]
+
+		glog.Info("Density test parameters: %v", dtest)
+
+		//if ppm==0, its a raw density test.
+		//otherwise, we continue launching n nodes per pod in threads till we meet the totalPods #.
+		if dtest.podsPerMinion == 0 {
+			//basic density tests
+			name := fmt.Sprintf("should allow starting %d pods per node", dtest.totalPods)
```

Comment:
I don't think this change is what we would like to do. The purpose of this test is to test how the performance looks in clusters with different sizes. So if we're testing a cluster with 2 nodes and a cluster with 100 nodes, the number of pods that we would like to start should be different (what I mean is that starting 30 pods on 100 nodes cluster doesn't expose Kubelets on heavy load at all).

I will prepare a PR for it.

BTW, I've just discovered that It() with function that takes parameters from "outside" world doesn't work (i.e. there are races). Will try to fix that too.